### PR TITLE
refactor: rust vectordb API stabilization of the Connection trait

### PR DESCRIPTION
**Pull request by westonpace**
This is the start of a more comprehensive refactor and stabilization of the Rust API.  The `Connection` trait is cleaned up to not require `lance` and to match the `Connection` trait in other APIs.  In addition, the concrete implementation `Database` is hidden.

BREAKING CHANGE: The struct `crate::connection::Database` is now gone.  Several examples opened a connection using `Database::connect` or `Database::connect_with_params`.  Users should now use `vectordb::connect`.

BREAKING CHANGE: The `connect`, `create_table`, and `open_table` methods now all return a builder object.  This means that a call like `conn.open_table(..., opt1, opt2)` will now become `conn.open_table(...).opt1(opt1).opt2(opt2).execute()`  In addition, the structure of options has changed slightly.  However, no options capability has been removed.

### CODE REVIEW

**Comment by westonpace** [comment]
Some questions for reviewers about preferred style:

 * Most operations take some kind of optional options object.  In `lance`, and previous iterations of `vectordb` we have used `Option<...>` or overloads (e.g. `connect` vs `connect_with_params`).  With the new APIs I am favoring a single method that always requires an options object that has a `Default` impl.  In other words:

```
// Yes
let table = conn.open_table("foo", OpenTableOptions::default());
// No
let table = conn.open_table("foo", None);
let table = conn.open_table("foo");
```

* Should options objects be all-pub structs?  This is simpler to start with but potentially less future-proof as things change.  For example:

```
// Current style (simple)
let opts = OpenTableOptions {
  index_cache_size: 64,
  ..Default::default()
};
// Builder style (more future proof)
let opts = OpenTableOptions::new().index_cache_size(64);
```

* Should options objects be passed in by value or reference?

Currently, I'm always having them passed by value (but be easily `Clone`d).  My thinking is that options eventually end up needing something that can't be freely cloned (e.g. a `Vec` / `Arc`) and taking by value can avoid an extra `Clone`.

* Use `&str/String` or `AsRef<str>/Into<String>`?

Most APIs are using `&str/String` right now.  Should we try and use `AsRef/Into` where easily done?

* "params" vs. "options"

There seems to be some places we use "params" and some places we use "options".  Should we consolidate on one word choice?

**Comment by wjones127** [comment]
> Most operations take some kind of optional options object.

I'd prefer either a method with 1 options object or a fluent builder API ([like in object-store](https://docs.rs/object_store/latest/object_store/aws/struct.AmazonS3Builder.html)).

> Should options objects be all-pub structs?

I guess the future-proof-ness of builders is we can deprecate certain methods? Or adjust the behavior / defaults?

> Should options objects be passed in by value or reference?

Value is good. 

> Most APIs are using &str/String right now. Should we try and use AsRef/Into where easily done?

We can really only do the latter with builders, right?

I'm kind of thinking maybe we should make the API use builders, given the advantages.

**Comment by westonpace** [comment]
> I guess the future-proof-ness of builders is we can deprecate certain methods? Or adjust the behavior / defaults?

Adjusting the behavior / defaults is what I was thinking.  For example, if we decide that `index_cache_size` is a tuple `(u32, u64)` of num_entries / num_bytes in the future then we can keep the existing builder method which would not just set the first part of the tuple.

> I'm kind of thinking maybe we should make the API use builders, given the advantages.

Switched to builders.

File: rust/vectordb/src/connection.rs [hunk 1]
```diff
@@ -29,73 +29,288 @@ use snafu::prelude::*;
...
+    #[allow(dead_code)] // https://github.com/lancedb/lancedb/issues/992
+    /// The invalid/missing items are replaced by fill_value
+    Fill(f32),
+}
+
+impl Default for BadVectorHandling {
+    fn default() -> Self {
+        Self::Error
+    }
+}
```
Comment (wjones127) [inline rust/vectordb/src/connection.rs:83]: TIL: you can derive `Default` for enums if the default value is a unit variant:
```suggestion
#[derive(Clone, Debug, Default)]
enum BadVectorHandling {
    /// An error is returned
    #[default]
    Error,
    #[allow(dead_code)] // https://github.com/lancedb/lancedb/issues/992
    /// The offending row is droppped
    Drop,
    #[allow(dead_code)] // https://github.com/lancedb/lancedb/issues/992
    /// The invalid/missing items are replaced by fill_value
    Fill(f32),
}
```

File: rust/vectordb/src/table.rs [hunk 2]
```diff
@@ -85,6 +84,40 @@ pub struct OptimizeStats {
...
+    Append,
+    /// The existing table will be overwritten with the new data
+    Overwrite,
+}
+
+impl Default for AddDataMode {
+    fn default() -> Self {
+        Self::Append
+    }
+}
```
Comment (wjones127) [inline rust/vectordb/src/table.rs:111]:
```suggestion
#[derive(Debug, Clone, Default)]
pub enum AddDataMode {
    /// Rows will be appended to the table (the default)
    #[default]
    Append,
    /// The existing table will be overwritten with the new data
    Overwrite,
}
```

File: rust/vectordb/src/table.rs [hunk 3]
```diff
@@ -214,32 +264,44 @@ pub trait Table: std::fmt::Display + Send + Sync {
     ///
     /// ```no_run
     /// # use std::sync::Arc;
-    /// # use vectordb::connection::{Database, Connection};
```
Comment (wjones127) [inline rust/vectordb/src/table.rs:217]: Don't have to do this in this PR, but: If we feel like merge_insert is sufficiently tested, we could use [the technique described here to shorten this and other examples](https://rust-unofficial.github.io/patterns/idioms/rustdoc-init.html).

**Review by wjones127 — APPROVED** [review]
Some minor suggestions, but otherwise looks good!

File: rust/vectordb/src/table.rs (hunk 3 above)
Comment (westonpace) [reply rust/vectordb/src/table.rs:217]: Oh man, I had been looking for something like this a while ago.  This is a very cool idea.  I will wait for a follow-up though.  I want to revisit the examples in a little more detail once the refactor is done (and I need to update the rst).

**Pull request merged**